File: c++/leetcode107.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result;
        vector<struct TreeNode*> tmp;
        vector<struct TreeNode*> container;
        if (!root) return result;
        container.push_back(root);
        while(!container.empty())
        {
            vector<int> level;
            tmp.clear();
            for(int i=0; i<container.size(); i++)
            {
                level.push_back(container[i]->val);
                if(container[i]->left)
                {
                    tmp.push_back(container[i]->left);
                }
                if(container[i]->right)
                {
                    tmp.push_back(container[i]->right);
                }
            }
            result.push_back(level);
            container = tmp;
        }
        std:: reverse(result.begin(), result.end());
        return result;
    }
};

File: python/leetcode107.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []
        if not root:
            return result
        tmp = []
        container = []
        container.append(root)
        while container:
            tmp.clear()
            level = []
            for key in container:
                print("key: ", key.val)
                level.append(key.val)
                if key.left is not None:
                    print("key left", key.left.val)
                    tmp.append(key.left)
                if key.right is not None:
                    print("key right", key.right.val)
                    tmp.append(key.right)
            result.append(level)
            print(level)
            container = list(tmp)
        result.reverse()
        return result
            
        
